src/Storage/Device/S3.php

```diff
@@ -165,30 +170,148 @@ public function getPath(string $filename): string
+        
+        $data = \file_get_contents($source);
+        $this->headers['content-type'] = \mime_content_type($source);
+        $this->headers['content-md5'] = \base64_encode(md5($data, true));
+        $this->amzHeaders['x-amz-content-sha256'] = \hash('sha256', $data);
+        unset($this->amzHeaders['x-amz-acl']); // ACL header is not allowed in parts, only createMultipartUpload accepts this header.
+
+        $response = $this->call(self::METHOD_PUT, $uri, $data, [
+            'partNumber'=>$chunk,
+            'uploadId' => $uploadId
+        ]);
+
+        return $response->headers['etag'];
+    }
+
+    /**
+     * Complete Multipart Upload
+     * 
+     * @param string $path
+     * @param string $uploadId
+     * @param array $parts
+     * 
      * @return bool
      */
-    public function upload($source, $path): bool
+    protected function completeMultipartUpload(string $path, string $uploadId, array $parts): bool
     {
-        return $this->write($path, \file_get_contents($source), \mime_content_type($source));
+        $uri = $path !== '' ? '/' . \str_replace(['%2F', '%3F'], ['/', '?'], \rawurlencode($path)) : '/';
+
+        $body = '<CompleteMultipartUpload>';
+        foreach ($parts as $part) {
+            $body .= "<Part><ETag>{$part['etag']}</ETag><PartNumber>{$part['partNumber']}</PartNumber></Part>";
+        }
+        $body .= '</CompleteMultipartUpload>';
+
+        $this->amzHeaders['x-amz-content-sha256'] = \hash('sha256', $body);
+        $this->headers['content-md5'] = \base64_encode(md5($body, true));
+        $this->call(self::METHOD_POST, $uri, $body , ['uploadId' => $uploadId]);
+        return true;
     }
 
     /**
-     * Read file by given path.
+     * Abort Chunked Upload
+     * 
+     * @param string $path
+     * @param mixed $extra
+     * 
+     * @return bool
+     */
+    public function abort(string $path, mixed $extra = ''): bool
+    {
+        $uri = $path !== '' ? '/' . \str_replace(['%2F', '%3F'], ['/', '?'], \rawurlencode($path)) : '/';
+        unset($this->headers['content-type']);
+        $this->headers['content-md5'] = \base64_encode(md5('', true));
+        $this->call(self::METHOD_DELETE, $uri, '', ['uploadId' => $extra]);
+        return true;
+    }
+
+    /**
+     * Read file or part of file by given path, offset and length.
      *
      * @param string $path
+     * @param int offset
+     * @param int length
      * 
      * @throws \Exception
      *
      * @return string
      */
-    public function read(string $path): string
+    public function read(string $path, int $offset = 0, int $length = null): string
     {
+        unset($this->headers['content-type']);
+        $this->headers['content-md5'] = \base64_encode(md5('', true));
         $uri = ($path !== '') ? '/' . \str_replace('%2F', '/', \rawurlencode($path)) : '/';
+        if($length !== null) {
+            $end = $offset + $length - 1;
+            $this->headers['range'] = "bytes=$offset-$end";
```

Comment:
<img width="431" alt="Screenshot 2021-12-03 at 1 46 32 AM" src="https://user-images.githubusercontent.com/20852629/144508316-a19b5e67-2413-4c9d-8d54-bdc0fba0ba13.png">

@lohanidamodar  what is the intended effect here ? 

```php
$offset = 10;
$end = 6;

"bytes=10-6" 

or 

"bytes=4"
```

Reply:
It is case 1 we are sending the byte range we want to read